File: algorithms/[1306]跳跃游戏 III/solution.ts
// topics = ["深度优先遍历"]

function canReach(arr: number[], start: number): boolean {
  const n = arr.length;
  const visited = new Set<number>();

  const dfs = (i: number) => {
    visited.add(i);

    const pre = i - arr[i];
    const next = i + arr[i];

    if (!visited.has(pre) && 0 <= pre && pre < n) {
      dfs(pre);
    }

    if (!visited.has(next) && 0 <= next && next < n) {
      dfs(next);
    }
  };

  dfs(start);

  return [...visited].map(i => arr[i]).includes(0);
}

File: algorithms/[1306]跳跃游戏 III/solution.py
# topics = ["广度优先遍历"]

from typing import Deque, List, Set
from collections import deque


class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        """
        BFS
        time O(n), space O(n), n 为数组长度
        """
        n = len(arr)
        q: Deque[int] = deque()
        visited: Set[int] = set()

        q.append(start)
        while q:
            i = q.popleft()
            visited.add(i)

            pre, next = i - arr[i], i + arr[i]
            if 0 <= pre < n and pre not in visited:
                q.append(pre)
            if 0 <= next < n and next not in visited:
                q.append(next)

        for k, v in enumerate(arr):
            if v == 0 and k in visited:
                return True

        return False
